cpuid/src/common.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(target_arch = "x86")]
+use std::arch::x86::CpuidResult;
+#[cfg(target_arch = "x86_64")]
+use std::arch::x86_64::CpuidResult;
+
+#[cfg(target_arch = "x86")]
+use std::arch::x86::__get_cpuid_max;
+#[cfg(target_arch = "x86_64")]
+use std::arch::x86_64::__get_cpuid_max;
+
+#[cfg(target_arch = "x86")]
+use std::arch::x86::__cpuid_count;
+#[cfg(target_arch = "x86_64")]
+use std::arch::x86_64::__cpuid_count;
+
+use kvm::CpuId;
+use kvm_bindings::kvm_cpuid_entry2;
+
+const INTEL: &[u8; 12] = b"GenuineIntel";
+const AMD: &[u8; 12] = b"AuthenticAMD";
+
+const EXT_FUNCTION: u32 = 0x80000000;
+
+pub enum Error {
+    InvalidParameters(String),
+    NotSupported,
+    SizeLimitExceeded,
+}
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+fn get_cpuid(function: u32, count: u32) -> Result<CpuidResult, Error> {
+    // For x86 the host supports the `cpuid` instruction if SSE is enabled. Otherwise it's hard to check.
+    // TODO: replace with validation based on `has_cpuid()` when it becomes stable:
+    //  https://doc.rust-lang.org/core/arch/x86/fn.has_cpuid.html
+    #[cfg(target_arch = "x86")]
+        {
+            #[cfg(not(target_feature = "sse"))]
+                {
+                    return Err(Error::NotSupported);
+                }
+        }
+
+    // this is safe because the host supports the `cpuid` instruction
+    let max_function = unsafe { __get_cpuid_max(function & EXT_FUNCTION).0 };
+    if function > max_function {
+        return Err(Error::InvalidParameters(format!(
+            "Function not supported: 0x{:x}",
+            function
+        )));
+    }
+
+    // this is safe because the host supports the `cpuid` instruction
+    let entry = unsafe { __cpuid_count(function, count) };
+    if entry.eax == 0 && entry.ebx == 0 && entry.ecx == 0 && entry.edx == 0 {
+        return Err(Error::InvalidParameters(format!(
+            "Invalid count: {}",
+            count
+        )));
+    }
+
+    Ok(entry)
+}
+
+/// Extracts the CPU vendor id from leaf 0x0.
+///
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn get_vendor_id() -> Result<[u8; 12], Error> {
+    match get_cpuid(0, 0) {
+        Ok(vendor_entry) => {
+            let bytes: [u8; 12] = unsafe {
+                std::mem::transmute([vendor_entry.ebx, vendor_entry.edx, vendor_entry.ecx])
+            };
+            Ok(bytes)
+        }
+        Err(_e) => Err(Error::NotSupported),
+    }
+}
+
+/// Replaces the `cpuid` entries corresponding to `function` with the entries from the host's cpuid.
+///
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn emulate_host_cpuid_function(cpuid: &mut CpuId, function: u32) -> Result<(), Error> {
```

Comment:
I find the name of the function misleading. I wouldn't use emulate because in fact (as you state in the doc string), we are using the host cpuid and the kvm emulated cpuid. We aren't emulating any features ourselves.

Reply:
I find it a bit inefficient to copy all entries whenever you want to adjust one. I do not have another solution at the moment, but we can take this offline and do a bit of brainstorming.

Reply:
Yap, sounds good.

Reply:
I created the task #1019 for this